File: pickle/putty/pickle/ParamControl/src/BinTables.cpp
#include <vcl.h>
#pragma hdrstop

#include "ti.h"
#include "BinTables.h"

const int ValidSignature = 0xBADFACED;
#define VALID_MARK(SUB) if (Mark.AnsiCompare(#SUB) != 0) throw Exception("Invalid Mark " #SUB)

TPC_Application PC_Application;
TPC_Table      *PC_Table;
TPC_Relation   *PC_Relation;
TPC_Field      *PC_Field;
TPC_Enum       *PC_Enum;
TPC_IndexField *PC_KeyField;
TPC_IndexField *PC_OrderField;
TPC_IndexField *PC_BreakField;
TPC_IndexField *PC_ShowField;
TPC_Link       *PC_Link;
TPC_LinkPair   *PC_LinkPair;

static int ReadInt(FILE *In)
{
  int Data;
  fread(&Data, sizeof(int), 1, In);
  char *p = (char*)&Data;
  p[0]^=p[3]^=p[0]^=p[3];
  p[1]^=p[2]^=p[1]^=p[2];
  return Data;
}

static AnsiString ReadString(FILE *In)
{
  char Buffer[2048];
  int Length = ReadInt(In);
  memset(Buffer, 0, sizeof(Buffer));
  fread(Buffer, Length, 1, In);
  return AnsiString(Buffer);
}

static void LoadApplication(FILE *In)
{
  int Signature = ReadInt(In);
  if (Signature != ValidSignature)
    throw Exception("Invalid Configuration Binary File");
  AnsiString Mark = ReadString(In);
  VALID_MARK(APPLICATION);
  PC_Application.name = ReadString(In);
  int p = PC_Application.name.Pos(" DESCR:");
  if (p > 0)
  {
    PC_Application.descr = PC_Application.name.SubString(p+7, PC_Application.name.Length());
    PC_Application.name.Delete(p, PC_Application.name.Length());
    p = PC_Application.descr.Pos(" VERSION:");
    if (p > 0)
    {
      PC_Application.version = PC_Application.descr.SubString(p+9, PC_Application.descr.Length());
      PC_Application.descr.Delete(p, PC_Application.descr.Length());
    }
  }
  PC_Application.registry = ReadString(In);
  #define READ(SUB) \
    PC_Application.no##SUB##s = ReadInt(In); \
    if (PC_Application.no##SUB##s) \
      PC_##SUB = new TPC_##SUB [PC_Application.no##SUB##s]
  READ(Table);
  READ(Relation);
  READ(Field);
  READ(Enum);
  READ(KeyField);
  READ(OrderField);
  READ(ShowField);
  READ(BreakField);
  READ(Link);
  READ(LinkPair);
  #undef READ
}

static void TPC_LoadTables(FILE *In)
{
  for (int i=0; i<PC_Application.noTables; i++)
  {
    PC_Table[i].name = ReadString(In);
    PC_Table[i].descr = ReadString(In);
    #define READ(SUB) PC_Table[i].##SUB = ReadInt(In);
    READ(noFields);
    READ(offsetFields);
    READ(noLinks);
    READ(offsetLinks);
    READ(noKeyFields);
    READ(offsetKeyFields);
    READ(noOrderFields);
    READ(offsetOrderFields);
    READ(noShowFields);
    READ(offsetShowFields);
    READ(noBreakFields);
    READ(offsetBreakFields);
    READ(viewOnly);
    #undef READ
  }
}

static void TPC_LoadRelations(FILE *In)
{
  for (int i=0; i<PC_Application.noRelations; i++)
  {
    PC_Relation[i].name = ReadString(In);
    PC_Relation[i].descr = ReadString(In);
    #define READ(SUB) PC_Relation[i].##SUB = ReadInt(In);
    READ(fromTable);
    READ(noFromFields);
    READ(toTable);
    READ(noToFields);
    READ(offsetFields);
    READ(offsetFromLink);
    READ(offsetToLink);
    #undef READ
  }
}

static void TPC_LoadFields(FILE *In)
{
  for (int i=0; i<PC_Application.noFields; i++)
  {
    PC_Field[i].name = ReadString(In);
    #define READ(SUB) PC_Field[i].##SUB = ReadInt(In);
    READ(type);
    READ(length);
    READ(precision);
    READ(scale);
    READ(offset);
    READ(paddedLength);
    READ(uppercase);
    READ(null);
    READ(noEnums);
    READ(offsetEnums);
    #undef READ
  }
}

static void TPC_LoadEnums(FILE *In)
{
  for (int i=0; i<PC_Application.noEnums; i++)
  {
    PC_Enum[i].name = ReadString(In);
    #define READ(SUB) PC_Enum[i].##SUB = ReadInt(In);
    READ(value);
    #undef READ
  }
}

static void TPC_LoadKeyFields(FILE *In)
{
  for (int i=0; i<PC_Application.noKeyFields; i++)
  {
    #define READ(SUB) PC_KeyField[i].##SUB = ReadInt(In);
    READ(index);
    #undef READ
  }
}

static void TPC_LoadOrderFields(FILE *In)
{
  for (int i=0; i<PC_Application.noOrderFields; i++)
  {
    #define READ(SUB) PC_OrderField[i].##SUB = ReadInt(In);
    READ(index);
    #undef READ
  }
}

static void TPC_LoadShowFields(FILE *In)
{
  for (int i=0; i<PC_Application.noShowFields; i++)
  {
    #define READ(SUB) PC_ShowField[i].##SUB = ReadInt(In);
    READ(index);
    #undef READ
  }
}

static void TPC_LoadBreakFields(FILE *In)
{
  for (int i=0; i<PC_Application.noBreakFields; i++)
  {
    #define READ(SUB) PC_BreakField[i].##SUB = ReadInt(In);
    READ(index);
    #undef READ
  }
}

static void TPC_LoadLinks(FILE *In)
{
  for (int i=0; i<PC_Application.noLinks; i++)
  {
    #define READ(SUB) PC_Link[i].##SUB = ReadInt(In);
    READ(tableNo);
    READ(noLinkPairs);
    READ(offsetLinkPairs);
    #undef READ
  }
}

static void TPC_LoadLinkPairs(FILE *In)
{
  for (int i=0; i<PC_Application.noLinkPairs; i++)
  {
    #define READ(SUB) PC_LinkPair[i].##SUB = ReadInt(In);
    READ(fromNo);
    READ(toNo);
    #undef READ
  }
}

void TPC_LoadParameters(char* BinFilename)
{
  FILE *In = fopen(BinFilename, "rb");
  if (In == 0) return;
  LoadApplication(In);
  AnsiString Mark = ReadString(In);  VALID_MARK(TABLES);
  if (PC_Table)      TPC_LoadTables(In);
  Mark = ReadString(In);   VALID_MARK(RELATIONS);
  if (PC_Relation)   TPC_LoadRelations(In);
  Mark = ReadString(In);   VALID_MARK(FIELDS);
  if (PC_Field)      TPC_LoadFields(In);
  Mark = ReadString(In);   VALID_MARK(ENUMS);
  if (PC_Enum)       TPC_LoadEnums(In);
  Mark = ReadString(In);   VALID_MARK(KEYFIELDS);
  if (PC_KeyField)   TPC_LoadKeyFields(In);
  Mark = ReadString(In);   VALID_MARK(ORDERFIELDS);
  if (PC_OrderField) TPC_LoadOrderFields(In);
  Mark = ReadString(In);   VALID_MARK(SHOWFIELDS);
  if (PC_ShowField) TPC_LoadShowFields(In);
  Mark = ReadString(In);   VALID_MARK(BREAKFIELDS);
  if (PC_BreakField) TPC_LoadBreakFields(In);
  Mark = ReadString(In);   VALID_MARK(LINKS);
  if (PC_Link)       TPC_LoadLinks(In);
  Mark = ReadString(In);   VALID_MARK(LINKPAIRS);
  if (PC_LinkPair)   TPC_LoadLinkPairs(In);
  Mark = ReadString(In);   VALID_MARK(ENDOFDATA);
}

void TPC_FreeParameters()
{
  if (PC_Table)       delete [] PC_Table;
  if (PC_Relation)    delete [] PC_Relation;
  if (PC_Field)       delete [] PC_Field;
  if (PC_Enum)        delete [] PC_Enum;
  if (PC_KeyField)    delete [] PC_KeyField;
  if (PC_OrderField)  delete [] PC_OrderField;
  if (PC_BreakField)  delete [] PC_BreakField;
  if (PC_Link)        delete [] PC_Link;
  if (PC_LinkPair)    delete [] PC_LinkPair;
}

File: picktester/ParamControl/BinTables.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace bbd.ParamControl
{
  class BinTables
  {
    static internal Encoding encoding = Encoding.GetEncoding("ISO-8859-1");
    internal const uint ValidSignature = 0xBADFACED;
    static public TPCApplication PCApplication = new TPCApplication();
    static internal TPCTable[] PCTables;
    static internal TPCRelation[] PCRelations;
    static internal TPCField[] PCFields;
    static internal TPCEnum[] PCEnums;
    static internal TPCIndexField[] PCKeyFields;
    static internal TPCIndexField[] PCOrderFields;
    static internal TPCIndexField[] PCBreakFields;
    static internal TPCIndexField[] PCShowFields;
    static internal TPCLink[] PCLinks;
    static internal TPCLinkPair[] PCLinkPairs;
    static void ValidateMark(BinaryReader In, String Mark)
    {
      string work = ReadString(In);
      if (Mark != work)
        throw new Exception("Invalid Mark " + Mark + " " + work + " returned.");
    }
    internal static void Allocate<T>(BinaryReader In, out int no, ref T[] array)
    {
      no = ReadInt(In);
      if (no > 0)
        array = new T[no];
      else
        array = null;
    }
    static void LoadApplication(BinaryReader In)
    {
      int Signature = ReadInt(In);
      if ((uint)Signature != ValidSignature)
        throw new Exception("Invalid Configuration Binary File");
      ValidateMark(In, "APPLICATION");
      PCApplication.name = ReadString(In);
      PCApplication.descr = ReadString(In);
      PCApplication.version = ReadString(In);
      PCApplication.server = ReadBytes(In);
      PCApplication.user = ReadBytes(In);
      PCApplication.password = ReadBytes(In);
      PCApplication.registry = ReadString(In);
      Allocate(In, out PCApplication.noTables, ref PCTables);
      Allocate(In, out PCApplication.noRelations, ref PCRelations);
      Allocate(In, out PCApplication.noFields, ref PCFields);
      Allocate(In, out PCApplication.noEnums, ref PCEnums);
      Allocate(In, out PCApplication.noKeyFields, ref PCKeyFields);
      Allocate(In, out PCApplication.noOrderFields, ref PCOrderFields);
      Allocate(In, out PCApplication.noShowFields, ref PCShowFields);
      Allocate(In, out PCApplication.noBreakFields, ref PCBreakFields);
      Allocate(In, out PCApplication.noLinks, ref PCLinks);
      Allocate(In, out PCApplication.noLinkPairs, ref PCLinkPairs);
      PCApplication.validateInit = ReadString(In);
      PCApplication.validateAll = ReadString(In);
      PCApplication.validateOther = ReadString(In);
    }
    static internal void LoadTables(BinaryReader In)
    {
      ValidateMark(In, "TABLES");
      for (int i = 0; i < PCApplication.noTables; i++)
      {
        PCTables[i].index = i;
        PCTables[i].name = ReadString(In);
        PCTables[i].descr = ReadString(In);
        PCTables[i].noFields = ReadInt(In);
        PCTables[i].offsetFields = ReadInt(In);
        PCTables[i].noLinks = ReadInt(In);
        PCTables[i].offsetLinks = ReadInt(In);
        PCTables[i].noKeyFields = ReadInt(In);
        PCTables[i].offsetKeyFields = ReadInt(In);
        PCTables[i].noOrderFields = ReadInt(In);
        PCTables[i].offsetOrderFields = ReadInt(In);
        PCTables[i].noShowFields = ReadInt(In);
        PCTables[i].offsetShowFields = ReadInt(In);
        PCTables[i].noBreakFields = ReadInt(In);
        PCTables[i].offsetBreakFields = ReadInt(In);
        PCTables[i].viewOnly = ReadInt(In);
        PCTables[i].validate = ReadString(In);
      }
    }
    static internal void LoadRelations(BinaryReader In)
    {
      ValidateMark(In, "RELATIONS");
      for (int i = 0; i < PCApplication.noRelations; i++)
      {
        PCRelations[i].index = i;
        PCRelations[i].name = ReadString(In);
        PCRelations[i].descr = ReadString(In);
        PCRelations[i].fromTable = ReadInt(In);
        PCRelations[i].noFromFields = ReadInt(In);
        PCRelations[i].toTable = ReadInt(In);
        PCRelations[i].noToFields = ReadInt(In);
        PCRelations[i].offsetFields = ReadInt(In);
        PCRelations[i].offsetFromLink = ReadInt(In);
        PCRelations[i].offsetToLink = ReadInt(In);
        PCRelations[i].validate = ReadString(In);
      }
    }
    static internal void LoadFields(BinaryReader In)
    {
      ValidateMark(In, "FIELDS");
      for (int i = 0; i < PCApplication.noFields; i++)
      {
        PCFields[i].name = ReadString(In);
        PCFields[i].type = ReadInt(In);
        PCFields[i].length = ReadInt(In);
        PCFields[i].precision = ReadInt(In);
        PCFields[i].scale = ReadInt(In);
        PCFields[i].offset = ReadInt(In);
        PCFields[i].paddedLength = ReadInt(In);
        PCFields[i].uppercase = ReadInt(In);
        PCFields[i].isNull = ReadInt(In);
        PCFields[i].noEnums = ReadInt(In);
        PCFields[i].offsetEnums = ReadInt(In);
      }
    }
    static internal void LoadEnums(BinaryReader In)
    {
      ValidateMark(In, "ENUMS");
      for (int i = 0; i < PCApplication.noEnums; i++)
      {
        PCEnums[i].name = ReadString(In);
        PCEnums[i].value = ReadInt(In);
      }
    }
    static internal void LoadKeyFields(BinaryReader In)
    {
      ValidateMark(In, "KEYFIELDS");
      for (int i = 0; i < PCApplication.noKeyFields; i++)
      {
        PCKeyFields[i].index = ReadInt(In);
      }
    }
    static internal void LoadOrderFields(BinaryReader In)
    {
      ValidateMark(In, "ORDERFIELDS");
      for (int i = 0; i < PCApplication.noOrderFields; i++)
      {
        PCOrderFields[i].index = ReadInt(In);
      }
    }
    static internal void LoadShowFields(BinaryReader In)
    {
      ValidateMark(In, "SHOWFIELDS");
      for (int i = 0; i < PCApplication.noShowFields; i++)
      {
        PCShowFields[i].index = ReadInt(In);
      }
    }
    static internal void LoadBreakFields(BinaryReader In)
    {
      ValidateMark(In, "BREAKFIELDS");
      for (int i = 0; i < PCApplication.noBreakFields; i++)
      {
        PCBreakFields[i].index = ReadInt(In);
      }
    }
    static internal void LoadLinks(BinaryReader In)
    {
      ValidateMark(In, "LINKS");
      for (int i = 0; i < PCApplication.noLinks; i++)
      {
        PCLinks[i].tableNo = ReadInt(In);
        PCLinks[i].noLinkPairs = ReadInt(In);
        PCLinks[i].offsetLinkPairs = ReadInt(In);
      }
    }
    static internal void LoadLinkPairs(BinaryReader In)
    {
      ValidateMark(In, "LINKPAIRS");
      for (int i = 0; i < PCApplication.noLinkPairs; i++)
      {
        PCLinkPairs[i].fromNo = ReadInt(In);
        PCLinkPairs[i].toNo = ReadInt(In);
      }
    }
    static internal int ReadInt(BinaryReader In)
    {
      return
        (In.ReadByte() << 24
        | In.ReadByte() << 16
        | In.ReadByte() << 8
        | In.ReadByte()
        );
    }
    static internal String ReadString(BinaryReader In)
    {
      int length = ReadInt(In);
      byte[] chars = new byte[length];
      chars = In.ReadBytes(length);
      string result = encoding.GetString(chars);
      return result.TrimEnd('\0');
    }
    static internal byte[] ReadBytes(BinaryReader In)
    {
      int length = ReadInt(In);
      byte[] chars = new byte[length];
      chars = In.ReadBytes(length);
      return chars;
    }
    static public void LoadParameters(String BinFilename)
    {
      using (BinaryReader In = new BinaryReader(new FileStream(BinFilename, FileMode.Open)))
      {
        LoadApplication(In);
        LoadTables(In);
        LoadRelations(In);
        LoadFields(In);
        LoadEnums(In);
        LoadKeyFields(In);
        LoadOrderFields(In);
        LoadShowFields(In);
        LoadBreakFields(In);
        LoadLinks(In);
        LoadLinkPairs(In);
        ValidateMark(In, "ENDOFDATA");
      }
    }
    static public void FreeParameters()
    {
      PCTables = null;
      PCRelations = null;
      PCFields = null;
      PCEnums = null;
      PCKeyFields = null;
      PCOrderFields = null;
      PCBreakFields = null;
      PCLinks = null;
      PCLinkPairs = null;
    }
    static public bool HasValidation()
    {
      if (PCApplication.validateAll.Length > 0) return true;
      if (PCApplication.validateOther.Length > 0) return true;
      foreach (TPCTable table in PCTables) if (table.validate.Length > 0) return true; 
      return false;
    }
  }
}
